### PR TITLE
feat: #9725 add shift property to modify register value in modbus input plugin

### PR DESCRIPTION
**Pull request by TimurDela**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

adds new 'shift' property to add a constant to the register value see issue #9725 

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

Minor changes:
* remove unused functions (type conversions without scale)
* reorder some checks (name was called in error message before checking it existed)
* adds new 'shift' property to add a constant to the register value see issue #9725

### CODE REVIEW

File: plugins/inputs/modbus/configuration_register.go [hunk 1]
```diff
@@ -36,21 +36,22 @@ const sampleConfigPartPerRegister = `
...
+    { name = "frequency",    byte_order = "AB",   data_type = "UFIXED", scale=0.1,   shift=0.0, address = [7]},
+    { name = "power",        byte_order = "ABCD", data_type = "UFIXED", scale=0.1,   shift=0.0, address = [3,4]},
   ]
   input_registers = [
-    { name = "tank_level",   byte_order = "AB",   data_type = "INT16",   scale=1.0,     address = [0]},
-    { name = "tank_ph",      byte_order = "AB",   data_type = "INT16",   scale=1.0,     address = [1]},
-    { name = "pump1_speed",  byte_order = "ABCD", data_type = "INT32",   scale=1.0,     address = [3,4]},
+    { name = "tank_level",   byte_order = "AB",   data_type = "INT16",   scale=1.0,	shift=0.0, address = [0]},
+    { name = "tank_ph",      byte_order = "AB",   data_type = "INT16",   scale=1.0,	shift=0.0, address = [1]},
+    { name = "pump1_speed",  byte_order = "ABCD", data_type = "INT32",   scale=1.0,	shift=0.0, address = [3,4]},
```
Comment (srebhan) [inline plugins/inputs/modbus/configuration_register.go:54]: See above.

File: plugins/inputs/modbus/configuration_register.go [hunk 2]
```diff
@@ -60,6 +61,7 @@ type fieldDefinition struct {
 	ByteOrder   string   `toml:"byte_order"`
 	DataType    string   `toml:"data_type"`
 	Scale       float64  `toml:"scale"`
+	Shift       float64  `toml:"shift"`
```
Comment (srebhan) [inline plugins/inputs/modbus/configuration_register.go:64]:
```suggestion
	Offset       float64  `toml:"offset"`
```

File: plugins/inputs/modbus/configuration_register.go [hunk 3]
```diff
@@ -216,7 +218,7 @@ func (c *ConfigurationOriginal) validateFieldDefinitions(fieldDefs []fieldDefini
 
 			// check scale
 			if item.Scale == 0.0 {
-				return fmt.Errorf("invalid scale '%f' in '%s' - '%s'", item.Scale, registerType, item.Name)
+				// return fmt.Errorf("invalid scale '%f' in '%s' - '%s'", item.Scale, registerType, item.Name)
```
Comment (srebhan) [inline plugins/inputs/modbus/configuration_register.go:221]: Ummmm... why?

File: plugins/inputs/modbus/configuration_request.go [hunk 4]
```diff
@@ -92,6 +92,7 @@ type requestFieldDefinition struct {
 	Name        string  `toml:"name"`
 	InputType   string  `toml:"type"`
 	Scale       float64 `toml:"scale"`
+	Shift       float64 `toml:"shift"`
```
Comment (srebhan) [inline plugins/inputs/modbus/configuration_request.go:95]:
```suggestion
	Offset       float64 `toml:"offset"`
```

File: plugins/inputs/modbus/configuration_request.go [hunk 5]
```diff
@@ -144,6 +145,11 @@ func (c *ConfigurationPerRequest) Check() error {
 
 		// Check the fields
 		for fidx, f := range def.Fields {
+			// Name is mandatory
+			if f.Name == "" {
+				return fmt.Errorf("empty field name in request for slave %d", def.SlaveID)
+			}
+
```
Comment (srebhan) [inline plugins/inputs/modbus/configuration_request.go:152]: No! You are changing the logic here. Currently, the name is _only_ mandatory if you do **not** omit the field, i.e. if you specify `omit = true` the field does not need a name. With your change, this is not true anymore.
Anyway, this change is completely unrelated to the topic of your PR. If there is a reason to change this, please submit an additional PR.

File: plugins/inputs/modbus/modbus_test.go [hunk 6]
```diff
@@ -152,6 +162,27 @@ func TestHoldingRegisters(t *testing.T) {
 			write:     []byte{0x08, 0x98},
 			read:      float64(220),
 		},
+		{
+			name:      "register0_ab_float32_shift",
+			address:   []uint16{0},
+			quantity:  1,
+			byteOrder: "AB",
+			dataType:  "FLOAT32",
+			shift:     10.,
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus_test.go:171]: Please be expressive
```suggestion
			shift:     10.0,
```

File: plugins/inputs/modbus/modbus_test.go [hunk 7]
```diff
@@ -233,7 +264,16 @@ func TestHoldingRegisters(t *testing.T) {
 			read:      float64(1234567890.9876544),
 		},
 		{
-			name:      "register0_ab_float",
+			name:      "register0_ab_float_noscale_noShift",
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus_test.go:267]: Please be consistent, either
```suggestion
			name:      "register0_ab_float_noScale_noShift",
```
or 
```golang
			name:      "register0_ab_float_noscale_noshift",
```

File: plugins/inputs/modbus/modbus_test.go [hunk 8]
```diff
@@ -292,15 +383,45 @@ func TestHoldingRegisters(t *testing.T) {
...
 			address:   []uint16{10},
 			quantity:  1,
 			byteOrder: "AB",
 			dataType:  "UINT16",
 			scale:     1,
+			shift:     12.0,
 			write:     []byte{0xAB, 0xCD},
-			read:      uint16(43981),
+			read:      uint16(43993),
 		},
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus_test.go:425]: Please do not remove existing tests, but rather add new ones. They are there for a reason (in this case to check identity-scaling)!

File: plugins/inputs/modbus/modbus_test.go [hunk 9]
```diff
@@ -402,15 +583,45 @@ func TestHoldingRegisters(t *testing.T) {
...
 			address:   []uint16{90, 91},
 			quantity:  2,
 			byteOrder: "ABCD",
 			dataType:  "UINT32",
 			scale:     1,
+			shift:     -2864434398,
 			write:     []byte{0xAA, 0xBB, 0xCC, 0xDD},
-			read:      uint32(2864434397),
+			read:      uint32(0),
 		},
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus_test.go:625]: See above. Please add a test instead of modifying one.

File: plugins/inputs/modbus/type_conversions16.go [hunk 10]
```diff
@@ -84,54 +29,54 @@ func determineConverterI16Scale(outType, byteOrder string, scale float64) (field
...
 		return func(b []byte) interface{} {
 			in := int16(tohost(b))
-			return int64(float64(in) * scale)
+			return int64(float64(in)*scale + shift)
 		}, nil
 	case "UINT64":
 		return func(b []byte) interface{} {
 			in := int16(tohost(b))
-			return uint64(float64(in) * scale)
+			return uint64(math.Max(0, float64(in)*scale+shift))
```
Comment (srebhan) [inline plugins/inputs/modbus/type_conversions16.go:42]: Why do you do that here, but not for other operations? Anyway, this changes the current behavior and you should have a good reason for doing so...

File: plugins/inputs/modbus/type_conversions.go [hunk 11]
```diff
@@ -2,53 +2,27 @@ package modbus
...
-		return determineConverterU64Scale(outType, byteOrder, scale)
-	case "FLOAT32":
-		return determineConverterF32Scale(outType, byteOrder, scale)
-	case "FLOAT64":
-		return determineConverterF64Scale(outType, byteOrder, scale)
-	}
-	return nil, fmt.Errorf("invalid input data-type: %s", inType)
-}
-
-func determineConverterNoScale(inType, byteOrder, outType string) (fieldConverterFunc, error) {
```
Comment (srebhan) [inline plugins/inputs/modbus/type_conversions.go:34]: No. Your approach will **always** convert the input to a floating point number and, even worse, performs a multiplication before converting to the output type. This will lead to a loss of precision!!! Try
```golang
package main

import (
	"fmt"
)

func main() {
	a := ^uint64(0)
	fmt.Println(a)

	b := uint64(float64(a))
	fmt.Println(b)

	fmt.Println(a == b)
}
```

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks for contributing this PR @TimurDela!
I think we should name the new parameter `offset` instead of `shift`, but I'm open for discussion. 
However, your changes to the conversion are fundamentally flawed and should be reverted! You will end up with a potential loss of precision for _all_ integer types as you convert to `float` in an intermediate step! Furthermore, you should _not_ change existing tests, but rather add new ones. The best really is to literally **copy** existing tests (the whole `Test...` function in that case) and modify them in the copy.

File: plugins/inputs/modbus/README.md [hunk 12]
```diff
@@ -85,15 +85,16 @@ Registers via Modbus TCP or Modbus RTU/ASCII.
   ## data_type   - INT16, UINT16, INT32, UINT32, INT64, UINT64,
   ##               FLOAT32-IEEE, FLOAT64-IEEE (the IEEE 754 binary representation)
   ##               FLOAT32, FIXED, UFIXED (fixed-point representation on input)
-  ## scale       - the final numeric variable representation
+  ## scale       - the final numeric variable representation will be (scale × variable + shift)
+  ## shift       - the final numeric variable representation will be (scale × variable + shift)
```
[the remark this answers is not in the document]
Comment (TimurDela) [reply plugins/inputs/modbus/README.md:89]: I initially also called it offset but, in Modbus, the offset usually refers to a change to the address in the Modbus message and I think it can be confusing (there are already variables called offset in the code).
Also, if we are to use namings used in electronics, then `scale` should be changed to `gain` but this would make former configuration files incompatible.

File: plugins/inputs/modbus/configuration_register.go (hunk 3 above)
Comment (TimurDela) [reply plugins/inputs/modbus/configuration_register.go:221]: My understanding is that if the scale is not specified, go sets it to 0.0 but the user probably had 1.0 in mind. Is there another way to distinguish the case where the user actively wrote 0.0 or did not set the field at all?

File: plugins/inputs/modbus/configuration_request.go (hunk 5 above)
Comment (TimurDela) [reply plugins/inputs/modbus/configuration_request.go:152]: The problem is that searching for name duplicates requires the name to exist. Also, `omit` does not exist in the configuration_register syntax.

File: plugins/inputs/modbus/modbus_test.go (hunk 8 above)
Comment (TimurDela) [reply plugins/inputs/modbus/modbus_test.go:425]: None of the existing tests have been removed, Their names were changed to distinguish them from the added tests. It is a result of git diff that it looks like tests have been altered

File: plugins/inputs/modbus/type_conversions16.go (hunk 10 above)
Comment (TimurDela) [reply plugins/inputs/modbus/type_conversions16.go:42]: If the shift is negative and larger than scale × in, we get a negative number and the uint16 conversion leads to very weird results. This is not an issue for the signed data types.

File: plugins/inputs/modbus/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (TimurDela) [reply plugins/inputs/modbus/README.md:89]: Thanks for your feedback @srebhan 
Many of the things you pointed out in the review come from the fact that I had the *register* configuration in mind when I wrote the code, whereas some functions were written with the *request* configuration. I believe it is a problem that both these configurations have different behaviours.
For instance, 
* request has *omit* and *output* but register not
* register accepts "AB" byte order (as it should be for INT16) but request does not
* register expects *address* to be an array but request wants a single int
* other checks are also different

I believe this makes the code more difficult to maintain and makes the user's life more difficult. Do you want me to try to modify the code more so that both syntaxes converge?

[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/modbus/README.md:89]: No please don't. This is by intention as we need to keep backward-compatibility with old config-files. Therefore, the configuration schemes differ quite a bit. I don't think this complicates code as the _config parsing code_ is separated and the core operates on a common scheme. Furthermore, please make changes to the config _style_ a separate discussion (maybe in an issue)...

File: plugins/inputs/modbus/configuration_register.go [hunk 13]
```diff
@@ -36,21 +36,22 @@ const sampleConfigPartPerRegister = `
...
-    { name = "energy",       byte_order = "ABCD", data_type = "FIXED", scale=0.001, address = [5,6]},
-    { name = "current",      byte_order = "ABCD", data_type = "FIXED", scale=0.001, address = [1,2]},
-    { name = "frequency",    byte_order = "AB",   data_type = "UFIXED", scale=0.1,  address = [7]},
-    { name = "power",        byte_order = "ABCD", data_type = "UFIXED", scale=0.1,  address = [3,4]},
+    { name = "power_factor", byte_order = "AB",   data_type = "FIXED", 	scale=0.01,  shift=0.0, address = [8]},
+    { name = "voltage",      byte_order = "AB",   data_type = "FIXED", 	scale=0.1,   shift=0.0, address = [0]},
+    { name = "energy",       byte_order = "ABCD", data_type = "FIXED", 	scale=0.001, shift=0.0, address = [5,6]},
+    { name = "current",      byte_order = "ABCD", data_type = "FIXED", 	scale=0.001, shift=0.0, address = [1,2]},
+    { name = "frequency",    byte_order = "AB",   data_type = "UFIXED", scale=0.1,   shift=0.0, address = [7]},
+    { name = "power",        byte_order = "ABCD", data_type = "UFIXED", scale=0.1,   shift=0.0, address = [3,4]},
```
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/modbus/configuration_register.go]: > Maybe value_offset?

Let's keep shift then...

File: plugins/inputs/modbus/configuration_register.go (hunk 3 above)
Comment (srebhan) [reply plugins/inputs/modbus/configuration_register.go:221]: Well in the old style it was mandatory to specify `scale`. You can change that, but silently accepting `0.0` which will then throw away data is not the right way. Either handle that default in the `init...(` functions or keep the error message that prevents data loss.

File: plugins/inputs/modbus/configuration_request.go (hunk 5 above)
Comment (srebhan) [reply plugins/inputs/modbus/configuration_request.go:152]: Well first of all you are in `configuration_request.go`. Second, it is insane to force users to specify names for stuff that they throw away anyway. And last but not least. The omitted fields are completely ignored, never get checked for duplicate names and are dropped in `groupFieldsToRequests` so their name is _never_ used anywhere. Where do you see an issue with the current code?

File: plugins/inputs/modbus/type_conversions16.go (hunk 10 above)
Comment (srebhan) [reply plugins/inputs/modbus/type_conversions16.go:42]: Well my point is, you only do it for `uint64` but the same issue (overflow(/underflow)) occurs for all integer types. So either check all or none.

File: plugins/inputs/modbus/type_conversions.go (hunk 11 above)
Comment (srebhan) [reply plugins/inputs/modbus/type_conversions.go:34]: This is **only** true for the (legacy) register-based configuration but **not** for the request-based configuration. The point here is that this _exactly_ reproduced previous behavior for register-based but improves stuff for request-based!

**Comment by TimurDela** [comment]
> 

I understand your concern. The problem is that by adding a number to another, it is possible to get out of the range of the data type. For instance with math.MaxInt16 +1. When looking at all the line changes, it looks like a lot of changes but in fact, I mainly kept the same logic except that
* I use big.Float instead of float64 (to avoid loss of precision with UINT64)
* I check that numbers do not go out of the data type range which I think is necessary as soon as we allow for shift

**Comment by srebhan** [comment]
@TimurDela therefor I suggested to _always_ use `float64` when doing math operations avoiding the issue...

**Comment by srebhan** [comment]
Again, there are fundamental issues in your code such as the forced conversion to float with potential precision loss and the large code change with a much higher complexity and 400 additional lines of code. Given that this functionality can already be achieved, I want to urge you to radically simplify your code while keeping the current properties, i.e. no conversion by default (only if `scale` is specified).

**Comment by TimurDela** [comment]
Can I please ask you to test this bit of code
```go
package main

import (
	"fmt"
	"math/big"
)

func main() {
	a := ^uint64(0)
	fmt.Println(a)

	var f *big.Float = big.NewFloat(0.0)
	f = f.SetUint64(a)
	b, _ := f.Uint64()
	fmt.Println(b)

	fmt.Println(a == b)
}
```
I agree that the syntax is verbose and so super easy to read but as you see, converting to big.Float does not introduce precision loss.

**Comment by srebhan** [comment]
@TimurDela fair enough, my main point remains. Why should we add this complexity and 400 lines of code to add something that can already be done?
Therefore, I'm asking you to simplify the code. E.g. entering the "scaling path" if either `scale != 0.0` or `shift != 0.0`. This way you can achieve what you want without all the additional code changes...
